### ISSUE TITLE
Memory ownership across pipelines

### ISSUE THREAD

**Issue by tmds** [davidfowl/BedrockFramework#47]
Memory ownership is tied to a Pipe which means you need to copy the data if you take it out of the pipe. And similarly, you need to copy data to get it into the pipe.

It seems by design. But maybe we could eliminate these copies by having more advanced Memory ownership? Though there may be a usability cost we don't want to pay.

**Comment by tmds** [comment]
cc @davidfowl

**Comment by jkotalik** [comment]
There was discussion about adding PipeWriter.Append which would take an IMemoryOwner and claim ownership over a block, but I think it was scrapped due to usability issues as you mentioned. It was easy to get into patterns where the consumer would append memory and continue to use that memory after appending. @halter73 has more context as that method existed in Kestrel before.

I'd imagine we'd need to reimplement Pipe.cs in order to add this functionality as well, which is just another barrier.

**Comment by halter73** [comment]
This was in the original version of System.IO.Pipelines (called Append like @jkotalik said), but got ripped out in API review because of the usability issues. Kestrel merely had ideas of using this and wasn't relying on this functionality at the time. It would be cool if we could somehow add it back, but I'm not holding my breath.

https://github.com/dotnet/corefxlab/pull/980/files#diff-cbfd130e7c317f2e490d96537111af0bR176-R213

If you can't see the highlighted code, you can expand the diff for PipelineReaderWriter.cs.

**Comment by davidfowl** [comment]
I don’t think we should add it back, at least not as a first class thing. I’m not convinced that can be used properly after our experience using pipelines in real software.

Maybe a custom pipe implementation where the caller understood the contract but there’s nothing that can or will be done in this repo to implement cooperative memory management

**Comment by tmds** [comment]
Closing old issue.
